src/audio/volume/volume.c

```diff
@@ -402,19 +395,144 @@ static struct comp_dev *volume_new(const struct comp_driver *drv,
 #endif
 	default:
 		comp_err(dev, "volume_new(): invalid ramp type %d", vol->ramp);
-		goto cd_fail;
+		rfree(cd);
+		return -EINVAL;
 	}
 
 	reset_state(cd);
-	comp_info(dev, "vol->initial_ramp = %d, vol->ramp = %d, vol->min_value = %d, vol->max_value = %d",
-		  vol->initial_ramp, vol->ramp,
-		  vol->min_value, vol->max_value);
+
+	return 0;
+}
+
+/* CONFIG_IPC_MAJOR_3 */
+#elif CONFIG_IPC_MAJOR_4
+static int set_volume_ipc4(struct vol_data *cd, uint32_t const channel,
+			   uint32_t const target_volume,
+			   uint32_t const curve_type,
+			   uint64_t const curve_duration)
+{
+	/* update target volume in peak_regs */
+	cd->peak_regs.target_volume_[channel] = target_volume;
+	/* update peak meter in peak_regs */
+	cd->peak_regs.peak_meter_[channel] = 0;
+
+	/* init target volume */
+	cd->tvolume[channel] = target_volume;
+	/* init ramp start volume*/
+	cd->rvolume[channel] = 0;
+	/* init muted volume */
+	cd->mvolume[channel] = 0;
+	/* set muted as false*/
+	cd->muted[channel] = false;
+#if CONFIG_COMP_VOLUME_WINDOWS_FADE
+	/* ATM there is support for the same ramp for all channels */
+	if (curve_type == IPC4_AUDIO_CURVE_TYPE_WINDOWS_FADE) {
+		cd->ramp_type = SOF_VOLUME_WINDOWS_FADE;
+		cd->ramp_func =  &volume_windows_fade_ramp;
+	}
+#endif
+	return 0;
+}
+
+static int init_volume_ipc4(struct comp_dev *dev,
+			    struct comp_ipc_config *config, void *spec)
+{
+	struct ipc4_peak_volume_module_cfg *vol = spec;
+	struct vol_data *cd = comp_get_drvdata(dev);
+	uint32_t channels_count;
+	uint8_t channel_cfg;
+	uint8_t channel;
+
+	dcache_invalidate_region(spec, sizeof(*vol));
+	mailbox_hostbox_read(&cd->base, sizeof(cd->base), 0, sizeof(cd->base));
+
+	channels_count = cd->base.audio_fmt.channels_count;
+
+	for (channel = 0; channel < channels_count ; channel++) {
+		if (vol->config[0].channel_id == IPC4_ALL_CHANNELS_MASK)
+			channel_cfg = 0;
+		else
+			channel_cfg = channel;
+
+		/* In IPC4 driver sends volume in Q1.31 format. It is converted
+		 * into Q1.23 format to be processed by firmware.
+		 */
+		vol->config[channel].target_volume >>= 8;
+
+		set_volume_ipc4(cd, channel,
+				vol->config[channel_cfg].target_volume,
+				vol->config[channel_cfg].curve_type,
+				vol->config[channel_cfg].curve_duration);
+	}
+
+	/* In IPC4 driver sends curve_duration in hundred of ns - it should be
+	 * converted into ms value required by firmware
+	 */
+	cd->initial_ramp  = vol->config[0].curve_duration / 10000;
```

Comment:
for my education, the comment and the calculation don't seem to add up ir ns->ms should divide by 1e6 isnt it? so why 1000 above?

Reply:
the IPC4 unit is somewhat strange ns/100